lib/vagrant/plugin/state_file.rb

```diff
@@ -91,9 +92,10 @@ def sources
 
       # This saves the state back into the state file.
       def save!
-        @path.open("w+") do |f|
-          f.write(JSON.dump(@data))
-        end
+        tmpfile = Tempfile.new(@path.basename.to_s, @path.dirname.to_s)
+        tmpfile.write(JSON.dump(@data))
+        tmpfile.close
+        File.rename(tmpfile.path, @path)
```

Comment:
Re 2: The tempfile is created in the same directory as the target
Re 3: I'll be honest and say this is my first Ruby patch, but I have a strong suspicion it can't be removed by GC because the local variable holds a strong reference across the function scope

But yes, this is just an implementation of a general pattern for "replace file atomically" which is harder on Windows indeed.  I'm an upstream develope for GLib where for `g_file_set_contents()`  we do https://git.gnome.org/browse/glib/tree/glib/gfileutils.c#n1217

It's the sort of thing I'd expect Ruby to have in the standard library, but if it exists I couldn't find it.